### DISCUSSION TITLE
Use the graph codegen

### DISCUSSION THREAD

**Discussion by DevLazio** [sreeise/graph-rs-sdk#339] [Q&A] [answered]
Hi !
I noticed there was a few endpoints that were missing (such as [/admin/serviceAnnouncement/issues](https://docs.microsoft.com/en-us/graph/api/serviceannouncement-list-issues?view=graph-rest-1.0&tabs=http) or [/admin/serviceAnnouncement/healthOverviews](https://docs.microsoft.com/en-us/graph/api/serviceannouncement-list-healthoverviews?view=graph-rest-1.0&tabs=http) ) so I wanted to try to implement them. 
But it looks like it most of the code should be generated via graph-codegen and I have no clue how to do that.

Would it be possible for you @sreeise to explains the basics steps to use it ?

**Comment by sreeise** [comment, marked as answer]
Hey,

Sure thing. This is going to be a quick overview of what you run to generate the code and then I will go into more detail. The high level overview of it works for the most part on the APIs that are already included but for new ones there is a little bit more work to be done. Ill try to explain things as best I can and then you can ask me any questions you may have since this may get a little complicated.


Since the /admin path is an API not yet added to this project, you will first want to add `Admin` to the `ResourceIdentity` Enum which would just be where Admin is in alphabetical order (at least I try to keep it that way).

https://github.com/sreeise/graph-rs/blob/6d75eb9616813d64f54d3606a8d5d59ecd3dbf50/graph-core/src/resource/resource_identity.rs#L20-L21

Next, you can generate the code by running the following code. I always just create a test.rs file in the examples folder since /examples/test.rs is set to be ignored by the gitignore file. The code to generate the /admin path will look like this:

```rust
use graph_codegen::macros::OpenApiParser;
use graph_codegen::openapi::OpenApi;
use graph_codegen::parser::client_resource::ResourceParsingInfo;
use graph_core::resource::ResourceIdentity;

fn main() {
    OpenApi::write(ResourceParsingInfo {
        modifier_name: Some("admin".into()),
        path: "/admin".to_string(),
        resource_identity: ResourceIdentity::Admin,
        parent_resource_info: None,
    });
}
```

This will create the src/admin directory along with the API macros in a request.rs file and its mod declarations in a mod.rs file inside the src/admin directory. 

Lastly you will want to do two things:

1. First add `pub mod admin;` to the list of mod declarations in the src/lib.rs file. Also meant to be in alphabetical order except for `pub mod client` which must be above all other mod declarations in order for the macros in src/client to be used throughout the project.
2. Then run `cargo fmt` so that you can actually read the newly generated code.

For any API thats already been generated before that is pretty much all you would need to do for this to work, with little to no extra effort for the most part.

Heres where the tricky part comes in. There are specific settings that has to be added to the codegen for each newly added API. If you run the code above it will generate the admin API in a way that does not compile correctly yet and really doesnt follow the way this project sets up APIs that follow a single path such /admin. You will probably see mutiple structs and their implementations. These do actually have something to do with the admin APIs and all start with the /admin path. This happens because OpenApi has an [Operations Object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#operation-object) that specifies an operation id, that is used in this project to generate the struct names and also the method names for individual requests. This would be fine if we were in an object oriented language that has classes and inheritance since that is how the operation id is meant to be read and used. 

For example, consider the operation id `reports.reportRoot.GetReportRoot` which is the original operation id for the Reports api GET request to /reports. If we were to leave this operation id as is, the code would be generated like so:

reports.reportRoot.GetReportRoot will be broken into three separate parts by splitting it using `.`  

* reports
* reportRoot
* GetReportRoot

We identify the method name for the GET endpoint by finding the last part of the operation id which always comes after the last `.` and is in Pascal Casing unlike the rest of the operation id. In normal object oriented languages you would take the first part, `Reports`, and set that as the parent class or possibly even an an interface, and then the `ReportRoot` would be a concrete child class with API methods for GET report root and POST report root. If you look in the reports directory, you will see the top two macros have the method names `get_report_root` and `update_report_root`.

I am not saying that this is how Microsoft does their code generation in C# or other OOP languages and its really just an example to illustrate. Microsoft has a weird complicated way of converting their APIs from XML to OpenApi specifications so its all kinda fudged up to begin with if your trying to parse their OpenApi configs.

Since Rust is not object oriented or at least not specifically designed to be even though some libraries do try to do a type of OOP, it becomes a little difficult to have a bunch of structs somehow linking each other.  Yes there is probably an elegant way to generate the APIs  using traits but for now I havnt had time to really look at this and so I decided to keep it simple. 

So how does the report root requests get changed to be included in the reports impl (there is no reports root struct in the reports directory). Its pretty simple, the settings in codegen tell it what operation id to use when we are generating reports APIs and find an operation id for the report root. 

https://github.com/sreeise/graph-rs/blob/6d75eb9616813d64f54d3606a8d5d59ecd3dbf50/graph-codegen/src/parser/settings/target_map_modifier.rs#L1148-L1163

So in order to generate the admin APIs correctly you would need to make sure that the operation ids and operation maps (the operation id without the method name) are set to remove any extra structs by adding a mapping like you see above for reports. There are also a few other settings such as ignore certain paths or generating a whole new api directory for paths that might be used by mutiple top level resources. Such as how /groups, /me, and /users have a calendars path, but we only generate one calendar api directory that all three of these use to cut down on code. 

If all of this seems confusing or overly complicated, then yeah your right. With the recent addition of actual models that follow the OpenApi specification, this is one area that I wanted to clean up and make the process more automated. But I just haven't had the time to really get into this lately. I can try to explain what settings need to be updated and how that all works but I figured this was enough for now. I also should probably explain the `ResourceParsingInfo` struct a little better and that it still needs some optimizing.

**Reply by sreeise** [reply]
Also going to add more too this soon. But I wanted to mention that the settings in target_map_modifiers.rs do not currently work when generating the code. I found the bug a few days ago and I am working on a fix. I should have mentioned that earlier so I apologize if you tried to generate code and are wondering why those settings are not working.

**Reply by sreeise** [reply]
Created a ticket for getting a better how to on generating code https://github.com/sreeise/graph-rs/issues/340
This should help move the process along faster.

**Reply by DevLazio** [reply]
Thanks a lot, that's exactly what I was looking for !

**Reply by sreeise** [reply]
Master now has the fix for target_map_modifier.rs and I am going to be working on #340 soon.

**Reply by sreeise** [reply]
Been looking at the admin API some more and I think this is one of those situations where it would be better to split this into three new APIs in the project which would be

- Admin
- ServiceAnnouncement
- HealthOverviews

ServiceAnnouncement and HealthOverviews are both part of the paths for the Admin API and these two end up having the same method names on some paths which can cause issues with generating the API without having to manually fix things. So from start to finish here is the settings I would add to the codegen and what I would run to generate the three of these.


So the first step would be to make the changes for ResourceIdentity by adding those above and then make the settings changes the the codegen files.
A commit of those changes is here: https://github.com/sreeise/graph-rs/compare/admin_api_settings?expand=1


Quick rundown of the settings:

- `ClientLinkSettings`: Creates the methods that link between APIs such as:
```rust
    impl<'a, Client> AdminRequest<'a, Client>
where
    Client: graph_http::RequestClient,
{
    pub fn service_announcement(&self) -> ServiceAnnouncementRequest<'a, Client> {
        self.client
            .request
            .extend_path(&[self.client.ident().as_ref()]);
        self.client.set_ident(ResourceIdentity::ServiceAnnouncement);
        ServiceAnnouncementRequest::new(self.client)
    }
}
```

Then you would be able to do this:
```rust
let client = Graph::new("");

client
  .v1()
  .admin()
  .service_announcement()
```
- `get_path_filters`: This method just tells us to ignore any paths that we don't want to have generated. So since we are generating ServiceAnnouncment API's we can just tell our code generator to ignore those paths when generating the Admin APIs. The `ClientLinkSettings` will make sure that the Admin client has a method for linking between the two.


Next you can run the codegen. You will need to generate the APIs in the order:
1. HealthOverviews
2. ServiceAnnouncement
3. Admin

This is just because of the way Admin depends on ServiceAnnouncments which depends on HealthOverviews. Meaning they import their crate. Run each method separately, and after each run 
1. add the pub mod declaration to src/libr.s for that api
2. run cargo fmt

Running codegen:
```rust
fn write_service_announcement() {
    OpenApi::write(ResourceParsingInfo {
        modifier_name: Some("serviceAnnouncement".into()),
        path: "/admin/serviceAnnouncement".to_string(),
        resource_identity: ResourceIdentity::ServiceAnnouncement,
        trim_path_start: "/admin".to_string(),
    });
}

fn write_health_overviews() {
    OpenApi::write(ResourceParsingInfo {
        modifier_name: Some("healthOverviews".into()),
        path: "/admin/serviceAnnouncement/healthOverviews".to_string(),
        resource_identity: ResourceIdentity::HealthOverviews,
        trim_path_start: Some("/admin/serviceAnnouncement".to_string())
    });
}

fn admin_api() {
    OpenApi::write(ResourceParsingInfo {
        modifier_name: Some("admin".into()),
        path: "/admin".to_string(),
        resource_identity: ResourceIdentity::Admin,
        trim_path_start: None,
    });
}
```

Lastly, don't forget to add the admin method to be able to use it with the client. This:
```rust
    pub fn admin(&self) -> AdminRequest<'a, Client> {
        self.client.set_ident(ResourceIdentity::Admin);
        AdminRequest::new(self.client)
    }

```
would go here right above `app_catalogs`: https://github.com/sreeise/graph-rs/blob/a3bb9bdcb202f499daeee30d76784301416c934a/src/client/graph.rs#L284

**Reply by sreeise** [reply]
Feel free to run through that and set up a PR if you want. If you run into any issues just let me know as well or if you want to just open a PR for it to discuss issues as well.

**Reply by DevLazio** [reply]
I was able to generate these API I use them after a few fixes before your patch for  target_map_modifier.rs  but I wasn't able to work on that since. I'll see if I can do that soon. 
Thanks for your help !

**Reply by DevLazio** [reply]
There is no "trim_path_start" field in the "ResourceParsingInfo" struct but there is one in the "ParentResourceInfo".
I ended up doing the trim by hand.

**Reply by sreeise** [reply]
Yeah sorry about that. I made changes that I pushed to master and then changed my example I gave here without telling you that it was updated. I will try to look at your PR here in the next couple of days.  

https://github.com/sreeise/graph-rs/blob/fbdafbe54140ac5f20c085f8c778b1610ceaa400/graph-codegen/src/parser/client_resource.rs#L5-L23
